Function to request locked user list

davidreneuw opened the pull request.

davidreneuw committed:
🚧 Built skeleton for request but I believe there is something wrong with the credential

davidreneuw committed:
Merge remote-tracking branch 'origin/dr/locked-users' into dr/locked-users

davidreneuw committed:
🚧 Minor refactoring

davidreneuw committed:
🚧 Implementation of locked users in admin tools user page

davidreneuw commented:
Added UI implementation for locked users. They will now have a lock emoji next to their name in the users table in the users admin tool:
![image](https://github.com/ssc-sp/datahub-portal/assets/56747050/2a0ea096-0d4a-4348-bd03-d06bd2a0d103)
I am planning on adding a button on the page to export a list of locked users to csv/txt files, is that a good idea?

I am still getting a client secret exception though, so currently commented the code that calls the function out and put a sample list of emails:
![image](https://github.com/ssc-sp/datahub-portal/assets/56747050/a64b8751-bf7f-4280-b8eb-c0ec700cf21a)
![image](https://github.com/ssc-sp/datahub-portal/assets/56747050/796651a3-023a-4488-a5df-ab8484cf48b2)

Ill have to look more into the function client secret issue. 
@yjmrobert thoughts on UI implementation and export button?

Also reiterating, I'm open to heavy criticism

davidreneuw committed:
🚧 Changing to match MSGraph API

yjmrobert commented:
Looks good! 

What if we put the lock icons in their own column so we can sort them or even a checkbox to only show locked users? Up to you!

Another idea would be a copy to clipboard instead of download CSV but either implementation feels appropriate

davidreneuw committed:
🚧 UI implementation of function, fixed the issue with function secret

davidreneuw committed:
✔️ Edited function to intersect with SP group members.

davidreneuw committed:
✔️ Refactored code to implement users status as a service. Added copy to clipboard button

davidreneuw commented:
I rewrote the code to implement user status as a service so that its easily available in the future.
Added copy to clipboard button as well as a question mark icon for users not present in MS Graph:
![image](https://github.com/ssc-sp/datahub-portal/assets/56747050/0f9ef9e8-9c23-4f7b-9213-c4d1ead5f6bb)

Function now also only returns the intersection of locked users and service principal's members.

Im not fully aware what the process is for publishing the function. In addition, we'll have to add a line to the appsettings

yjmrobert commented:
Nice, I'll take a look at the code.

For the deployment, the cicd should automatically deploy it! 

I'll message you re: the config value

davidreneuw committed:
✔️ Minor refactoring of the config variable name and fix on Status column sorting

davidreneuw committed:
✔️ Config refactor

davidreneuw committed:
✔️ Moved service file to proper location, added copy to clipboard buttons for all environment users and all tenant locked users with tooltips and a loading indicator

davidreneuw committed:
Merge branch 'develop' into dr/locked-users

davidreneuw commented:
Added copy to clipboard buttons with tooltips:
![image](https://github.com/ssc-sp/datahub-portal/assets/56747050/d710abe3-8a2e-4bba-be08-850e2d37fb08)
![image](https://github.com/ssc-sp/datahub-portal/assets/56747050/b20f32b7-14d2-46f7-85f8-8346d0106221)

Added loading indicator when resources arent loaded yet:
![image](https://github.com/ssc-sp/datahub-portal/assets/56747050/caed5952-964e-4384-b757-dafe997b1069)

Moved files around to follow infra/app structure better

yjmrobert commented on Portal/src/Datahub.Functions/GetUsersStatus.cs:
```diff
@@ -0,0 +1,103 @@
+            {
+                allUsers.Add(User.Mail);
+            }
+            _logger.LogInformation("Processed all MSGraph users");
```
I think we can omit this part. This will return all the users in the tenant, including ones that were never part of DataHub, and will likely flag an overreach of scope from the cloud team

yjmrobert replied:
To clarify, the SP group members would be "all" of our DataHub users

davidreneuw replied:
Fixed with most recent commit ✔️

yjmrobert commented on Portal/src/Datahub.Portal/Pages/Tools/Users/UsersPage.razor:
```diff
@@ -1,35 +1,65 @@
+            <MudButton Color="Color.Primary" OnClick="@CopyLockedUsersToClipboard">
+                @_lockedButton
+            </MudButton>
+        </MudTooltip>
```
Don't forget to wrap the text with the `@Localizer` 😄

davidreneuw replied:
Fixed with most recent commit ✔️

yjmrobert commented on Portal/src/Datahub.Portal/Pages/Tools/Users/UsersPage.razor:
```diff
@@ -1,35 +1,65 @@
+        if (!firstRender && !_hasLoaded)
+        {
+            _hasLoaded = _usersTable.loadStatus;
+            _lockedButton = "Copy locked users to clipboard";
+            _allButton = "Copy all users to clipboard";
```
Don't forget the `Localizer` ! 😄 

Any reason why these are held through variables instead of razor markup?

yjmrobert reviewed (changes requested):
Nice work, glad to see the pieces coming together! 

I'll reach out for some more info as I want to show you something I think you'll really enjoy 😁

iroelperez commented on Portal/src/Datahub.Infrastructure/Services/UsersStatusService.cs:
```diff
@@ -0,0 +1,46 @@
+            
+            do
+            {
+                using var client = _httpClientFactory.CreateClient();
```
I think you shouldn't dispose the httpClients (or check if this is what is recommended), I may be wrong but the httpClientFactory is who manages the lifecycle of the httpClients.

davidreneuw replied:
In most recent commit, moved the httpclient outside of the while loop and removed `using` keyword to let the factory manage the lifecycle

iroelperez commented on Portal/src/Datahub.Infrastructure/Services/UsersStatusService.cs:
```diff
@@ -0,0 +1,46 @@
+                var result = await client.GetAsync(url);
+                resultString = await result.Content.ReadAsStringAsync();
+            
+            } while (string.IsNullOrWhiteSpace(resultString) && numberOfRetries++ < maxNumberOfRetries);
```
This retry is probably because you get a throttling response from Azure (409), you can use a Retry Policy as we do with the Azure Management API in the Costing and Storage stats collection. I can show you if that is the case of the problem.

davidreneuw replied:
In most recent commit, added a proper retry policy, please let me know if it was implemented properly

iroelperez commented on Portal/src/Datahub.Portal/Pages/Tools/Users/UsersPage.razor:
```diff
@@ -1,35 +1,65 @@
+        if (!firstRender && !_hasLoaded)
+        {
+            _hasLoaded = _usersTable.loadStatus;
+            _lockedButton = "Copy locked users to clipboard";
```
Add the literals to the localization and replace the code with Localizer["english_text"] so we can localize them to French.

davidreneuw replied:
Done in most recent commit ✔️

iroelperez commented on Portal/src/Datahub.Portal/Pages/Tools/Users/UsersTable.razor:
```diff
@@ -22,6 +37,19 @@
+            @switch(GetUserStatus(@context.User.Email))
+            {
+                case "locked":
+                    <MudIcon Icon="@Icons.Material.Filled.Lock" Title="Account locked"/>
```
Localization missing here.

davidreneuw replied:
Added in most recent commit ✔️

davidreneuw committed:
✔️ Added localization for all texts added, added a proper retry policy, cleaned up a lot of the code into functional, removed all MSGraph users api call

davidreneuw committed:
Merge branch 'develop' into dr/locked-users
